File: csharp/src/Vista.SDK/LocalIdBuilder.cs
using System.Text;
using Vista.SDK.Internal;

namespace Vista.SDK;

public sealed partial record class LocalIdBuilder : ILocalIdBuilder<LocalIdBuilder, LocalId>
{
    public static readonly string NamingRule = "dnv-v2";
    public static readonly CodebookName[] UsedCodebooks =
    [
        CodebookName.Quantity,
        CodebookName.Content,
        CodebookName.State,
        CodebookName.Command,
        CodebookName.FunctionalServices,
        CodebookName.MaintenanceCategory,
        CodebookName.ActivityType,
        CodebookName.Position,
        CodebookName.Detail,
    ];
    public VisVersion? VisVersion { get; private init; }

    public bool VerboseMode { get; private init; } = false;

    internal LocalIdItems Items { get; private init; }

    public GmodPath? PrimaryItem => Items.PrimaryItem;

    public GmodPath? SecondaryItem => Items.SecondaryItem;

    public MetadataTag? Quantity { get; private init; }

    public MetadataTag? Content { get; private init; }

    public MetadataTag? Calculation { get; private init; }

    public MetadataTag? State { get; private init; }

    public MetadataTag? Command { get; private init; }

    public MetadataTag? Type { get; private init; }

    public MetadataTag? Position { get; private init; }

    public MetadataTag? Detail { get; private init; }

    public LocalIdBuilder WithVisVersion(in string visVersion)
    {
        var localIdbuilder = TryWithVisVersion(visVersion, out var succeeded);
        if (!succeeded)
            throw new ArgumentException("Failed to parse VIS version", nameof(visVersion));

        return localIdbuilder;
    }

    public LocalIdBuilder WithVisVersion(in VisVersion version)
    {
        var localIdbuilder = TryWithVisVersion(version, out var succeeded);
        if (!succeeded)
            throw new ArgumentException(nameof(WithVisVersion));

        return localIdbuilder;
    }

    public LocalIdBuilder TryWithVisVersion(in VisVersion? visVersion) => TryWithVisVersion(visVersion, out _);

    public LocalIdBuilder TryWithVisVersion(in string? visVersionStr, out bool succeeded)
    {
        if (VisVersions.TryParse(visVersionStr, out VisVersion v) == true)
        {
            var localIdbuilder = TryWithVisVersion(v, out var succeededInner);
            succeeded = succeededInner;
            return localIdbuilder;
        }
        succeeded = false;
        return this;
    }

    public LocalIdBuilder TryWithVisVersion(in VisVersion? visVersion, out bool succeeded)
    {
        succeeded = true;
        return this with { VisVersion = visVersion };
    }

    public LocalIdBuilder WithoutVisVersion() => this with { VisVersion = null };

    public LocalIdBuilder WithVerboseMode(in bool verboseMode) => this with { VerboseMode = verboseMode };

    public LocalIdBuilder WithPrimaryItem(in GmodPath item)
    {
        var localIdbuilder = TryWithPrimaryItem(item, out var succeeded);
        if (!succeeded)
            throw new ArgumentException(nameof(WithPrimaryItem));

        return localIdbuilder;
    }

    public LocalIdBuilder TryWithPrimaryItem(in GmodPath? item) => TryWithPrimaryItem(item, out _);

    public LocalIdBuilder TryWithPrimaryItem(in GmodPath? item, out bool succeeded)
    {
        if (item is null)
        {
            succeeded = false;
            return this;
        }

        succeeded = true;
        return this with { Items = this.Items with { PrimaryItem = item } };
    }

    public LocalIdBuilder WithoutPrimaryItem() => this with { Items = this.Items with { PrimaryItem = null } };

    public LocalIdBuilder WithSecondaryItem(in GmodPath item)
    {
        var localIdBuilder = TryWithSecondaryItem(item, out var succeeded);
        if (!succeeded)
            throw new ArgumentException(nameof(WithSecondaryItem));

        return localIdBuilder;
    }

    public LocalIdBuilder TryWithSecondaryItem(in GmodPath? item) => TryWithSecondaryItem(in item, out _);

    public LocalIdBuilder TryWithSecondaryItem(in GmodPath? item, out bool succeeded)
    {
        if (item is null)
        {
            succeeded = false;
            return this;
        }

        succeeded = true;
        return this with { Items = this.Items with { SecondaryItem = item } };
    }

    public LocalIdBuilder WithoutSecondaryItem() => this with { Items = this.Items with { SecondaryItem = null } };

    public LocalIdBuilder WithMetadataTag(in MetadataTag metadataTag)
    {
        var localIdBuilder = TryWithMetadataTag(metadataTag, out var succeeded);
        if (!succeeded)
            throw new ArgumentException("invalid metadata codebook name: " + metadataTag.Name);
        return localIdBuilder;
    }

    public LocalIdBuilder TryWithMetadataTag(in MetadataTag? metadataTag) => TryWithMetadataTag(metadataTag, out _);

    public LocalIdBuilder TryWithMetadataTag(in MetadataTag? metadataTag, out bool succeeded)
    {
        if (metadataTag is null)
        {
            succeeded = false;
            return this;
        }

        switch (metadataTag.Value.Name)
        {
            case CodebookName.Quantity:
                succeeded = true;
                return WithQuantity(metadataTag.Value);
            case CodebookName.Content:
                succeeded = true;
                return WithContent(metadataTag.Value);
            case CodebookName.Calculation:
                succeeded = true;
                return WithCalculation(metadataTag.Value);
            case CodebookName.State:
                succeeded = true;
                return WithState(metadataTag.Value);
            case CodebookName.Command:
                succeeded = true;
                return WithCommand(metadataTag.Value);
            case CodebookName.Type:
                succeeded = true;
                return WithType(metadataTag.Value);
            case CodebookName.Position:
                succeeded = true;
                return WithPosition(metadataTag.Value);
            case CodebookName.Detail:
                succeeded = true;
                return WithDetail(metadataTag.Value);
            default:
                succeeded = false;
                return this;
        }
    }

    public LocalIdBuilder WithoutMetadataTag(in CodebookName name) =>
        name switch
        {
            CodebookName.Quantity => WithoutQuantity(),
            CodebookName.Content => WithoutContent(),
            CodebookName.Calculation => WithoutCalculation(),
            CodebookName.State => WithoutState(),
            CodebookName.Command => WithoutCommand(),
            CodebookName.Type => WithoutType(),
            CodebookName.Position => WithoutPosition(),
            CodebookName.Detail => WithoutDetail(),
            _ => this
        };

    public LocalIdBuilder WithoutQuantity() => this with { Quantity = null };

    public LocalIdBuilder WithoutContent() => this with { Content = null };

    public LocalIdBuilder WithoutCalculation() => this with { Calculation = null };

    public LocalIdBuilder WithoutState() => this with { State = null };

    public LocalIdBuilder WithoutCommand() => this with { Command = null };

    public LocalIdBuilder WithoutType() => this with { Type = null };

    public LocalIdBuilder WithoutPosition() => this with { Position = null };

    public LocalIdBuilder WithoutDetail() => this with { Detail = null };

    private LocalIdBuilder WithQuantity(in MetadataTag quantity) => this with { Quantity = quantity, };

    private LocalIdBuilder WithContent(in MetadataTag content) => this with { Content = content, };

    private LocalIdBuilder WithCalculation(in MetadataTag calculation) => this with { Calculation = calculation, };

    private LocalIdBuilder WithState(in MetadataTag state) => this with { State = state, };

    private LocalIdBuilder WithCommand(in MetadataTag command) => this with { Command = command, };

    private LocalIdBuilder WithType(in MetadataTag type) => this with { Type = type, };

    private LocalIdBuilder WithPosition(in MetadataTag position) => this with { Position = position, };

    private LocalIdBuilder WithDetail(in MetadataTag detail) => this with { Detail = detail, };

    public static LocalIdBuilder Create(VisVersion version) => new LocalIdBuilder().WithVisVersion(version);

    public LocalId Build()
    {
        if (IsEmpty)
            throw new InvalidOperationException("Cant build to LocalId from empty LocalIdBuilder");
        if (!IsValid)
            throw new InvalidOperationException("Cant build to LocalId from invalid LocalIdBuilder");

        return new LocalId(this);
    }

    public bool HasCustomTag =>
        (Quantity?.IsCustom ?? false)
        || (Calculation?.IsCustom ?? false)
        || (Content?.IsCustom ?? false)
        || (Position?.IsCustom ?? false)
        || (State?.IsCustom ?? false)
        || (Command?.IsCustom ?? false)
        || (Type?.IsCustom ?? false)
        || (Detail?.IsCustom ?? false);

    public bool IsValid =>
        VisVersion is not null
        && Items.PrimaryItem is not null
        && (
            Quantity is not null
            || Calculation is not null
            || Content is not null
            || Position is not null
            || State is not null
            || Command is not null
            || Type is not null
            || Detail is not null
        );

    public bool IsEmpty =>
        Items.PrimaryItem is null
        && Items.SecondaryItem is null
        && (
            Quantity is null
            && Calculation is null
            && Content is null
            && Position is null
            && State is null
            && Command is null
            && Type is null
            && Detail is null
        );

    public bool IsEmptyMetadata =>
        Quantity is null
        && Calculation is null
        && Content is null
        && Position is null
        && State is null
        && Command is null
        && Type is null
        && Detail is null;

    public IReadOnlyList<MetadataTag> MetadataTags =>
        new List<MetadataTag?>() { Quantity, Calculation, Content, Position, State, Command, Type, Detail }
            .Where(m => m is not null)
            .Cast<MetadataTag>()
            .ToArray();

    public bool Equals(LocalIdBuilder? other)
    {
        if (other is null)
            return false;

        if (VisVersion != other.VisVersion)
            throw new InvalidOperationException("Cant compare local IDs from different VIS versions");

        return PrimaryItem == other.PrimaryItem
            && SecondaryItem == other.SecondaryItem
            && Quantity == other.Quantity
            && Calculation == other.Calculation
            && Content == other.Content
            && Position == other.Position
            && State == other.State
            && Command == other.Command
            && Type == other.Type
            && Detail == other.Detail;
    }

    public sealed override int GetHashCode()
    {
        var hashCode = new HashCode();
        hashCode.Add(PrimaryItem);
        hashCode.Add(SecondaryItem);
        hashCode.Add(Quantity);
        hashCode.Add(Calculation);
        hashCode.Add(Content);
        hashCode.Add(Position);
        hashCode.Add(State);
        hashCode.Add(Command);
        hashCode.Add(Type);
        hashCode.Add(Detail);
        return hashCode.ToHashCode();
    }

    public void ToString(StringBuilder builder)
    {
        if (VisVersion is null)
            throw new InvalidOperationException("No VisVersion configured on LocalId");

        string namingRule = $"/{NamingRule}/";

        builder.Append(namingRule);

        builder.Append("vis-");
        VisVersion.Value.ToVersionString(builder);
        builder.Append('/');

        Items.Append(builder, VerboseMode);

        builder.Append("meta/");

        // NOTE: order of metadatatags matter,
        // should not be changed unless changed in the naming rule/standard
        builder.AppendMeta(Quantity);
        builder.AppendMeta(Content);
        builder.AppendMeta(Calculation);
        builder.AppendMeta(State);
        builder.AppendMeta(Command);
        builder.AppendMeta(Type);
        builder.AppendMeta(Position);
        builder.AppendMeta(Detail);

        if (builder[builder.Length - 1] == '/')
            builder.Remove(builder.Length - 1, 1);
    }

    public override string ToString()
    {
        using var lease = StringBuilderPool.Get();
        var builder = lease.Builder;

        ToString(builder);

        return lease.ToString();
    }
}

File: js/lib/LocalId.Builder.ts
import { CodebookName } from "./CodebookName";
import { Codebooks } from "./Codebooks";
import { Gmod } from "./Gmod";
import { GmodPath } from "./GmodPath";
import { ILocalIdBuilderGeneric } from "./ILocalIdBuilder";
import { LocalIdParsingErrorBuilder } from "./internal/LocalIdParsingErrorBuilder";
import { LocalId } from "./LocalId";
import { LocalIdItems } from "./LocalId.Items";
import { LocalIdParser } from "./LocalId.Parsing";
import { Locations } from "./Location";
import { MetadataTag } from "./MetadataTag";
import { ParsingState } from "./types/LocalId";
import { VisVersion } from "./VisVersion";

export class LocalIdBuilder
    implements ILocalIdBuilderGeneric<LocalIdBuilder, LocalId>
{
    public static readonly namingRule = "dnv-v2";
    public static readonly usedCodebooks = [
        CodebookName.Quantity,
        CodebookName.Content,
        CodebookName.Calculation,
        CodebookName.State,
        CodebookName.Command,
        CodebookName.Type,
        CodebookName.Position,
        CodebookName.Detail,
    ];

    public visVersion?: VisVersion;
    public verboseMode = false;
    private _items: LocalIdItems;

    public quantity?: MetadataTag;
    public content?: MetadataTag;
    public calculation?: MetadataTag;
    public state?: MetadataTag;
    public command?: MetadataTag;
    public type?: MetadataTag;
    public position?: MetadataTag;
    public detail?: MetadataTag;

    private constructor() {
        this._items = new LocalIdItems();
    }

    public static create(visVersion: VisVersion): LocalIdBuilder {
        return new LocalIdBuilder().withVisVersion(visVersion);
    }

    public clone() {
        return Object.assign(new LocalIdBuilder(), this);
    }

    public build(): LocalId {
        return new LocalId(this);
    }

    public get primaryItem() {
        return this._items?.primaryItem;
    }

    public get secondaryItem() {
        return this._items?.secondaryItem;
    }

    public get isValid(): boolean {
        return (
            !!this._items.primaryItem &&
            !!this.visVersion &&
            !this.isEmptyMetadata
        );
    }

    public get hasCustomTag(): boolean {
        return this.metadataTags.some((t) => t.isCustom);
    }

    public get isEmpty(): boolean {
        return (
            !this._items.primaryItem &&
            !this._items.secondaryItem &&
            this.isEmptyMetadata
        );
    }

    public get isEmptyMetadata(): boolean {
        return this.metadataTags.length === 0;
    }

    public get metadataTags(): MetadataTag[] {
        const metaDataList = [
            this.quantity,
            this.content,
            this.calculation,
            this.state,
            this.command,
            this.type,
            this.position,
            this.detail,
        ];
        return metaDataList.filter((m) => m) as MetadataTag[];
    }

    public validate(
        errorBuilder = new LocalIdParsingErrorBuilder()
    ): LocalIdParsingErrorBuilder {
        // Add validation for primaryItem
        if (!this.primaryItem) {
            errorBuilder.push(ParsingState.PrimaryItem);
        }
        if (this.isEmptyMetadata) {
            errorBuilder.push({
                type: ParsingState.EmptyState,
                message: "Missing any metadata tag",
            });
        }

        return errorBuilder;
    }

    public equals(other?: LocalIdBuilder, skipOther = false): boolean {
        if (!other) return false;
        if (this.visVersion !== other.visVersion)
            throw new Error(
                "Cant compare local IDs from different VisVersions"
            );

        return !!(
            (this.primaryItem?.equals(other.primaryItem) ?? true) &&
            (this.secondaryItem?.equals(other.secondaryItem) ?? true) &&
            (this.quantity?.equals(other.quantity) ?? true) &&
            (this.calculation?.equals(other.calculation) ?? true) &&
            (this.content?.equals(other.content) ?? true) &&
            (this.position?.equals(other.position) ?? true) &&
            (this.state?.equals(other.state) ?? true) &&
            (this.command?.equals(other.command) ?? true) &&
            (this.type?.equals(other.type) ?? true) &&
            (this.detail?.equals(other.detail) ?? true) &&
            (skipOther ? true : other.equals(this, true))
        );
    }

    public toString(builder: string[] = []): string {
        if (!this.visVersion)
            throw new Error("No VisVersion configured on LocalId");
        const namingRule = `/${LocalIdBuilder.namingRule}/`;

        builder.push(namingRule);

        builder.push("vis-");
        builder.push(this.visVersion);
        builder.push("/");

        this._items.append(builder, this.verboseMode);

        builder.push("meta/");
        this.quantity?.append(builder);
        this.content?.append(builder);
        this.calculation?.append(builder);
        this.state?.append(builder);
        this.command?.append(builder);
        this.type?.append(builder);
        this.position?.append(builder);
        this.detail?.append(builder);

        if (builder[builder.length - 1].endsWith("/")) {
            builder[builder.length - 1] = builder[builder.length - 1].slice(
                0,
                -1
            );
        }
        const result = builder.join("");

        if (!LocalIdParser.matchISOString(result))
            throw new Error("Invalid characters in LocalId");

        return result;
    }

    public getMetadataTag(name: CodebookName): MetadataTag | undefined {
        switch (name) {
            case CodebookName.Position:
                return this.position;
            case CodebookName.Calculation:
                return this.calculation;
            case CodebookName.Quantity:
                return this.quantity;
            case CodebookName.State:
                return this.state;
            case CodebookName.Content:
                return this.content;
            case CodebookName.Command:
                return this.command;
            case CodebookName.Type:
                return this.type;
            case CodebookName.Detail:
                return this.detail;
            default:
                throw new Error(
                    `Unknown or invalid metadata tag name: ${name}`
                );
        }
    }

    public withVisVersion(visVersion: VisVersion): LocalIdBuilder {
        return this.with((s) => (s.visVersion = visVersion));
    }

    public withoutVisVersion(): LocalIdBuilder {
        return this.with((s) => (s.visVersion = undefined));
    }

    public tryWithVisVersion(visVersion?: VisVersion): LocalIdBuilder {
        if (visVersion === undefined) return this;
        return this.withVisVersion(visVersion);
    }

    public withVerboseMode(verboseMode: boolean): LocalIdBuilder {
        return this.with((s) => (s.verboseMode = verboseMode));
    }

    public withoutVerboseMode(): LocalIdBuilder {
        return this.with((s) => (s.verboseMode = false));
    }

    public withPrimaryItem(item: GmodPath): LocalIdBuilder {
        return this.with((s) => (s._items.primaryItem = item));
    }

    public withoutPrimaryItem(): LocalIdBuilder {
        return this.with((s) => (s._items.primaryItem = undefined));
    }

    public tryWithPrimaryItem(item?: GmodPath): LocalIdBuilder {
        if (!item) return this;
        return this.withPrimaryItem(item);
    }

    public withSecondaryItem(item: GmodPath): LocalIdBuilder {
        return this.with((s) => (s._items.secondaryItem = item));
    }

    public withoutSecondaryItem(): LocalIdBuilder {
        return this.with((s) => (s._items.secondaryItem = undefined));
    }

    public tryWithSecondaryItem(item?: GmodPath): LocalIdBuilder {
        if (!item) return this;
        return this.withSecondaryItem(item);
    }

    public withQuantity(metadataTag: MetadataTag): LocalIdBuilder {
        return this.with((s) => (s.quantity = metadataTag));
    }

    public withCalculation(metadataTag: MetadataTag): LocalIdBuilder {
        return this.with((s) => (s.calculation = metadataTag));
    }

    public withContent(metadataTag: MetadataTag): LocalIdBuilder {
        return this.with((s) => (s.content = metadataTag));
    }

    public withPosition(metadataTag: MetadataTag): LocalIdBuilder {
        return this.with((s) => (s.position = metadataTag));
    }

    public withState(metadataTag: MetadataTag): LocalIdBuilder {
        return this.with((s) => (s.state = metadataTag));
    }

    public withCommand(metadataTag: MetadataTag): LocalIdBuilder {
        return this.with((s) => (s.command = metadataTag));
    }

    public withType(metadataTag: MetadataTag): LocalIdBuilder {
        return this.with((s) => (s.type = metadataTag));
    }

    public withDetail(metadataTag: MetadataTag): LocalIdBuilder {
        return this.with((s) => (s.detail = metadataTag));
    }

    public tryWithMetadataTag(metadataTag?: MetadataTag): LocalIdBuilder {
        if (!metadataTag) return this;
        return this.withMetadataTag(metadataTag);
    }

    public withMetadataTag(metadataTag: MetadataTag): LocalIdBuilder {
        switch (metadataTag.name) {
            case CodebookName.Position:
                return this.withPosition(metadataTag);
            case CodebookName.Calculation:
                return this.withCalculation(metadataTag);
            case CodebookName.Quantity:
                return this.withQuantity(metadataTag);
            case CodebookName.State:
                return this.withState(metadataTag);
            case CodebookName.Content:
                return this.withContent(metadataTag);
            case CodebookName.Command:
                return this.withCommand(metadataTag);
            case CodebookName.Type:
                return this.withType(metadataTag);
            case CodebookName.Detail:
                return this.withDetail(metadataTag);
            default:
                throw new Error(
                    `Unknown or invalid metadata tag name: ${metadataTag.name}`
                );
        }
    }

    public withoutMetadataTag(name: CodebookName): LocalIdBuilder {
        switch (name) {
            case CodebookName.Position:
                return this.withoutPosition();
            case CodebookName.Calculation:
                return this.withoutCalculation();
            case CodebookName.Quantity:
                return this.withoutQuantity();
            case CodebookName.State:
                return this.withoutState();
            case CodebookName.Content:
                return this.withoutContent();
            case CodebookName.Command:
                return this.withoutCommand();
            case CodebookName.Type:
                return this.withoutType();
            case CodebookName.Detail:
                return this.withoutDetail();
            default:
                throw new Error(
                    `Unknown or invalid metadata tag name: ${name}`
                );
        }
    }

    private withoutPosition() {
        return this.with((s) => (s.position = undefined));
    }

    private withoutQuantity() {
        return this.with((s) => (s.quantity = undefined));
    }
    private withoutContent() {
        return this.with((s) => (s.content = undefined));
    }
    private withoutCommand() {
        return this.with((s) => (s.command = undefined));
    }
    private withoutState() {
        return this.with((s) => (s.state = undefined));
    }
    private withoutCalculation() {
        return this.with((s) => (s.calculation = undefined));
    }
    private withoutType() {
        return this.with((s) => (s.type = undefined));
    }

    private withoutDetail() {
        return this.with((s) => (s.detail = undefined));
    }

    public with(u: { (state: LocalIdBuilder): void }): LocalIdBuilder {
        const n = this.clone();
        u && u(n);
        return n;
    }

    public static parse(
        localIdStr: string | undefined,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return LocalIdParser.parse(
            localIdStr,
            gmod,
            codebooks,
            locations,
            errorBuilder
        );
    }

    public static async parseAsync(
        localIdString: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return LocalIdParser.parseAsync(localIdString, errorBuilder);
    }

    public static tryParse(
        localIdStr: string | undefined,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return LocalIdParser.tryParse(
            localIdStr,
            gmod,
            codebooks,
            locations,
            errorBuilder
        );
    }

    public static async tryParseAsync(
        localIdString: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return LocalIdParser.tryParseAsync(localIdString, errorBuilder);
    }
}
